### ISSUE TITLE
No available storage method found" error when implementing custom storage driver

### ISSUE THREAD

**Issue by snawar92** [jupyterlite/jupyterlite#1500]
## Description
I'm trying to implement a custom storage driver for JupyterLite to interface with a Rails backend. However, I'm encountering a "No available storage method found" error when attempting to use this custom driver.

## Steps to Reproduce
1. Created a custom storage driver implementing the `Contents.IDrive` interface
2. Registered the custom driver with JupyterLite
3. Configured `jupyter-lite.json` to use the custom driver
4. Loaded JupyterLite in the browser

## Expected Behavior
JupyterLite should recognize and use the custom storage driver for content management.

## Actual Behavior
JupyterLite throws a "No available storage method found" error and fails to initialize properly.

## Configuration
Here's my `jupyter-lite.json` configuration:
"jupyter-config-data": {
    "appPlugins": [
          { "id": "rails-storage-plugin", "url": "./plugins/custom-storage-driver.js" }
        ],
    "contentsStorageDrivers": ["customStorage"],
    "contentsStorageName": "customStorage",
}

**Comment by manics** [comment]
Note there's a duplicate post on Discourse https://discourse.jupyter.org/t/no-available-storage-method-found-error-when-implementing-custom-storage-driver/29267

**Comment by snawar92** [comment]
> Note there's a duplicate post on Discourse https://discourse.jupyter.org/t/no-available-storage-method-found-error-when-implementing-custom-storage-driver/29267

@manics can you help on this ? i am trying to use my custom api instead of indexdb storage , this is the reason i am trying to configure contentsStorageDrivers

**Comment by yemaney** [comment]
Reached a similar issue. I also couldn't find in depth documentation for implementing / using custom drivers. where is the best place to look for this?

**Comment by sadath-12** [comment]
+1

**Comment by snawar92** [comment]
Code 1

    <script>
      // Custom customAsyncStorageWrapper driver
      const customAsyncStorageWrapper = {
        _driver: 'customAsyncStorageWrapper',
        _initStorage: function (options) {
          console.log("============Initializing storage");
          this.apiBaseUrl = '/api/notebooks'; // Set your API base URL here
          return Promise.resolve();
        },
        _support: function () {
          return fetch(this.apiBaseUrl).then(() => true).catch(() => false);
        },
        iterate: function (e, n) {
          var t = this,
              r = new Promise(function (n, r) {
                t.ready()
                  .then(function () {
                    return fetch(t.apiBaseUrl);
                  })
                  .then(response => response.json())
                  .then(data => {
                    let i = 1;
                    for (let item of data) {
                      let result = e(item.content, item.name, i++);
                      if (result !== undefined) {
                        n(result);
                        return;
                      }
                    }
                    n();
                  })
                  .catch(r);
              });
          return r;
        },
        getItem: function (e, n) {
          var t = this;
          e = encodeURIComponent(e);
          var r = new Promise(function (n, r) {
            t.ready()
              .then(function () {
                return fetch(`${t.apiBaseUrl}/${e}`);
              })
              .then(response => response.json())
              .then(data => n(data.content))
              .catch(r);
          });
          return r;
        },
        setItem: function (e, n, t) {
          var r = this;
          e = encodeURIComponent(e);
          var o = new Promise(function (t, o) {
            r.ready()
              .then(function () {
                console.log("Attempting to save item:", e);
                return fetch(r.apiBaseUrl, {
                  method: 'POST',
                  headers: {
                    'Content-Type': 'application/json',
                    'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]')?.content
                  },
                  body: JSON.stringify({ name: e, content: n })
                });
              })
              .then(response => {
                if (!response.ok) {
                  throw new Error(`HTTP error! status: ${response.status}`);
                }
                return response.json();
              })
              .then(data => {
                console.log("Save successful:", data);
                t(data);
              })
              .catch(error => {
                console.error("Error saving item:", error);
                o(error);
              });
          });
          return o;
        },
        removeItem: function (e, n) {
          var t = this;
          e = encodeURIComponent(e);
          var r = new Promise(function (n, r) {
            t.ready()
              .then(function () {
                return fetch(`${t.apiBaseUrl}/${e}`, {
                  method: 'DELETE',
                  headers: {
                    'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]').content
                  }
                });
              })
              .then(() => n())
              .catch(r);
          });
          return r;
        },
        clear: function (e) {
          var n = this,
              t = new Promise(function (e, t) {
                n.ready()
                  .then(function () {
                    return fetch(n.apiBaseUrl, {
                      method: 'DELETE',
                      headers: {
                        'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]').content
                      }
                    });
                  })
                  .then(() => e())
                  .catch(t);
              });
          return t;
        },
        length: function (e) {
          var n = this,
              t = new Promise(function (e, t) {
                n.ready()
                  .then(function () {
                    return fetch(n.apiBaseUrl);
                  })
                  .then(response => response.json())
                  .then(data => e(data.length))
                  .catch(t);
              });
          return t;
        },
        key: function (e, n) {
          var t = this,
              r = new Promise(function (n, r) {
                e < 0
                  ? n(null)
                  : t.ready()
                      .then(function () {
                        return fetch(t.apiBaseUrl);
                      })
                      .then(response => response.json())
                      .then(data => {
                        if (e >= data.length) {
                          n(null);
                        } else {
                          n(data[e].name);
                        }
                      })
                      .catch(r);
              });
          return r;
        },
        keys: function (e) {
          var n = this,
              t = new Promise(function (e, t) {
                n.ready()
                  .then(function () {
                    return fetch(n.apiBaseUrl);
                  })
                  .then(response => response.json())
                  .then(data => e(data.map(item => item.name)))
                  .catch(t);
              });
          return t;
        },
        dropInstance: function (e, n) {
          var t = this,
              r = new Promise(function (n, r) {
                t.ready()
                  .then(function () {
                    return fetch(`${t.apiBaseUrl}/drop`, {
                      method: 'POST',
                      headers: {
                        'Content-Type': 'application/json',
                        'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]').content
                      },
                      body: JSON.stringify(e)
                    });
                  })
                  .then(() => n())
                  .catch(r);
              });
          return r;
        }
      };

      // Register the custom driver with localforage
      localforage.defineDriver(customAsyncStorageWrapper).then(() => {
        console.log("Custom driver registered successfully");
      }).catch((error) => {
        console.error("Error registering custom driver:", error);
      });

      // Set the custom driver as the preferred driver
      localforage.setDriver('customAsyncStorageWrapper').then(() => {
        console.log("Custom driver set as preferred driver");
      }).catch((error) => {
        console.error("Error setting custom driver as preferred:", error);
      });
    </script>





Code 2

 <script id="jupyter-config-data"      type="application/json"      data-jupyter-lite-root=".."    >
      {
        "contentsManager": {
          "constructor": "CustomContentsAPI"
        }
      }
    </script>

    <script>
      // Custom Contents API implementation
      class CustomContentsAPI {
        constructor(options) {
          this.apiBaseUrl = '/api/notebooks'; // Your custom API endpoint
        }

        async fetch(path, init) {
          const url = `${this.apiBaseUrl}/${path}`;
          const response = await fetch(url, init);
          if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
          }
          return response;
        }

        async get(path, options) {
          const response = await this.fetch(path);
          const data = await response.json();
          return this._adaptServerResponse(data);
        }

        async save(path, options) {
          const { content, format } = options;
          const response = await this.fetch(path, {
            method: 'PUT',
            headers: {
              'Content-Type': 'application/json',
            },
            body: JSON.stringify({ content, format }),
          });
          const data = await response.json();
          return this._adaptServerResponse(data);
        }

        async delete(path) {
          await this.fetch(path, { method: 'DELETE' });
        }

        _adaptServerResponse(data) {
          return {
            name: data.name,
            path: data.path,
            last_modified: data.last_modified,
            created: data.created,
            content: data.content,
            format: data.format,
            mimetype: data.mimetype,
            size: data.size,
            writable: data.writable,
            type: data.type,
          };
        }
      }

      // Make CustomContentsAPI available globally
      window.CustomContentsAPI = CustomContentsAPI;

      (async function () {
        const { pathname, origin, search, hash } = window.location;
        if (pathname.match(/(index.html|\/?)$/) == null) {
          window.location.href = `${origin}${pathname}/${search}${hash}`;
          return;
        }
        await import(
          '../config-utils.js?_=46e77a8'
        );
      }.call(this));
    </script>




i also tried these 2 approaches , and it didnt worked for me can you help me on this ?

@manics @andeplane @jtpio @andeplane @katotetsuro @trungleduc @krassowski

**Comment by ockhamlabs** [comment]
Hello Folks! Curious if this is something you can help with?

**Comment by jtpio** [comment]
Thanks all.

There may be some confusion in the documentation about the use of custom drivers and custom drives.

> Created a custom storage driver implementing the `Contents.IDrive` interface

A custom drive can be implemented in a regular JupyterLab extension, and will be added to the left area, similar to the default file browser. Good examples of such custom drives are:
- https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access
- https://github.com/jupyterlab/jupyterlab-github/

So you could use such approach to fetch and save files from / to a remote location, by implementing the `Contents.IDrive` interface.

> Configured `jupyter-lite.json` to use the custom driver

JupyterLite stores users settings, and by default also the files, locally in the user browser. This can be customized too, but may be more limited that the `Contents.IDrive` approach: https://jupyterlite.readthedocs.io/en/stable/howto/configure/storage.html.

So to summarize: if you want to fetch files from a remote location it will likely be easier to implement your own `Contents.IDrive`.

**Comment by snawar92** [comment]
@jtpio Thanks for the clarification! Implementing the Contents.IDrive interface for custom drives seems like a great way to manage remote files. The examples you shared are really helpful. Appreciate the insights!
